MaxText/layers/pipeline.py

```diff
@@ -70,6 +83,13 @@ def init_states(self, inputs):
     shift = jnp.zeros((self.num_stages,) + inputs.shape[1:], dtype=inputs.dtype)
     shift = nn.with_logical_constraint(shift, ("activation_stage", "activation_batch", "activation_length", "activation_embed"),rules=self.config.logical_axis_rules,mesh=self.mesh)
 
+    # Prev outputs has the same shape of the output (and shift)
+    if self.config.pipeline_delay_activation_forwarding:
+      prev_outputs = jnp.zeros((self.num_stages,) + inputs.shape[1:], dtype=inputs.dtype)
```

Comment:
Disclaimer: I have not thought hard about this, but I wonder if we can consolidate some of these buffers. e.g. combine `shift` and `circ_storage` for the intermediate activations, move the inputs from `state_io` directly into the intermediate buffer so the first stage reads from the same place as the others, and just place the inputs at an offset to achieve delayed forwarding. Then we'd just have buffers for `stage_inputs` and `outputs`.

Reply:
This is a good idea! I would have to think about this more. I have thought about similar solutions and it is tricky to remove circular storage - currently circular storage lives all on stage 0 - whereas your solution will keep it sharded, we would need to insert the circular output at the correct stage so it lands on stage 0 when it is needed. We also want this to be non-blocking though, so it should be on stage 0 even before it is needed - this is tricky

Reply:
Ah, good point! Keeping circ_storage sharded could also unnecessarily increase comms.